code/GatewayFieldsFactory.php

```diff
@@ -247,69 +213,93 @@ public function getCompanyFields()
     }
 
     /**
+     * Clear all fields that are not required by the gateway. Does nothing if gateway is null
+     * @param $fields
+     * @param array $defaults
+     */
+    protected function cullForGateway(&$fields, $defaults = array())
+    {
+        if (!$this->gateway){
+            return;
+        }
+
+        $selected = array_merge($defaults, GatewayInfo::requiredFields($this->gateway));
+        foreach ($fields as $name => $field) {
+            if (!in_array($name, $selected)) {
+                unset($fields[$name]);
+            }
+        }
+    }
+
+    /**
      * Attempts to find a custom field name and/or prefix defined in rename.yml, otherwise returns the same input
      * that it was given
      *
-     * @param string|array  $defaultName The default name of the field
-     * @param null|bool     $skipGateway Used in recursion
+     * @param string|array $defaultName The default name of the field
+     * @param null|bool    $skipGateway Used in recursion
      *
      * @return string|array
      */
-    public function getFieldName($defaultName, $skipGateway = null) {
-        // batch support
+    public function getFieldName($defaultName, $skipGateway = NULL)
+    {
         if (is_array($defaultName)) {
-            $stack = array();
-            foreach ($defaultName as $name) {
-                $stack[] = $this->getFieldName($name, $skipGateway);
-            }
-
-            return $stack;
+            return $this->getFieldNames($defaultName);
         }
 
-        $renameMap = \Config::inst()->get('GatewayFieldsFactory', 'rename');
+        $renameMap = $this->getRenameMap();
 
-        if (!$renameMap || empty($renameMap)) {
-            return $defaultName;
+        if (!$renameMap) {
+            return FALSE;
```

Comment:
You're right, my mistake, this was a fail copy paste from the old code before I rewrote it >.>